common/signal_processing/src/butterworth.cpp

```diff
@@ -0,0 +1,295 @@
+  /**
+   * The left and right limits of the magnitudes satisfy the specs at the
+   * frequencies Ws and Wp Scipy.buttord gives left limit as the cut-off frequency whereas Matlab
+   * gives right limit. We keep left_lim as a definition and commented out.
+   * */
+
+  double right_lim = Ws * (std::pow((std::pow(10.0, As / 10.0) - 1.0), -1.0 / (2. * order)));
+  // double left_lim = Wp * (std::pow((std::pow(10.0, Ap / 10.0) - 1.0), -1.0 / (2. * order)));
+
+  setCuttoffFrequency(right_lim);
+}
+
+void ButterworthFilter::setOrder(int const & N) { order_ = N; }
+
+void ButterworthFilter::setCuttoffFrequency(const double & Wc) { cutoff_frequency_rad_sec = Wc; }
+
+/**
+ * @brief Sets the cut-off and sampling frequencies.
+ * @param fc [in] cut-off frequency in Hz.
+ * @param fs [in] sampling frequency in Hz.
+ * */
+void ButterworthFilter::setCuttoffFrequency(const double & fc, const double & fs)
+{
+  /*
+   * fc is the cut-off frequency in [Hz]
+   * fs is the sampling frequency in [Hz]
+   * */
+  if (fc >= fs / 2) {
+    print("Invalid argument : Cut-off frequency  fc must be less than fs/2 \n");
+    return;
+  }
+
+  cutoff_frequency_rad_sec = fc * 2.0 * M_PI;
+  sampling_frequency_hz = fs;
+}
+
+sOrderCutoff ButterworthFilter::getOrderCutOff() const
+{
+  return sOrderCutoff{order_, cutoff_frequency_rad_sec};
+}
+
+/**
+ * @brief Matlab equivalent : [b, a]  = butter(n, Wn, 's')
+ * */
+void ButterworthFilter::computeContinuousTimeTF(const bool & use_sampling_frequency)
+{
+  // First compute  the phase angles of the roots
+  computePhaseAngles();
+  computeContinuousTimeRoots(use_sampling_frequency);
+
+  continuous_time_denominator_ = poly(continuous_time_roots_);
+  continuous_time_numerator_ = std::pow(cutoff_frequency_rad_sec, order_);
+}
+
+void ButterworthFilter::computePhaseAngles()
+{
+  phase_angles_.resize(order_, 0.);
+  int k{1};
+
+  for (auto & x : phase_angles_) {
+    x = M_PI_2 + (M_PI * (2.0 * k - 1.0) / (2.0 * order_));
+    k++;
+  }
+}
+
+void ButterworthFilter::computeContinuousTimeRoots(const bool & use_sampling_frequency)
+{
+  continuous_time_roots_.resize(order_, {0.0, 0.0});
+  int k{};
+
+  if (use_sampling_frequency) {
+    double const & Fc = (sampling_frequency_hz / M_PI) *
+                        tan(cutoff_frequency_rad_sec / (sampling_frequency_hz * 2.0));
+
+    for (auto & x : continuous_time_roots_) {
+      x = {
+        std::cos(phase_angles_[k]) * Fc * 2.0 * M_PI, std::sin(phase_angles_[k]) * Fc * 2.0 * M_PI};
+      k++;
+    }
+
+  } else {
+    for (auto & x : continuous_time_roots_) {
+      x = {
+        cutoff_frequency_rad_sec * cos(phase_angles_[k]),
+        cutoff_frequency_rad_sec * sin(phase_angles_[k])};
+      k++;
+    }
+  }
+}
+std::vector<std::complex<double>> ButterworthFilter::poly(
+  std::vector<std::complex<double>> const & roots)
+{
+  std::vector<std::complex<double>> coefficients(roots.size() + 1, {0, 0});
+
+  int const n{static_cast<int>(roots.size())};
```

Comment:
You are using `int const` everywhere, but IMO `const int` looks organized since other packages use it.